File: next/ts/electron/esm/core/ipc/MetaBody.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { cacheGetter } from "../../helper/cacheGetter.js";
import { simpleDecoder } from "../../helper/encoding.js";
import { IPC_DATA_ENCODING } from "./const.js";
export class MetaBody {
    constructor(type, senderUid, data, streamId, receiverUid, metaId = simpleDecoder(crypto.getRandomValues(new Uint8Array(8)), "base64")) {
        Object.defineProperty(this, "type", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: type
        });
        Object.defineProperty(this, "senderUid", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: senderUid
        });
        Object.defineProperty(this, "data", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: data
        });
        Object.defineProperty(this, "streamId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: streamId
        });
        Object.defineProperty(this, "receiverUid", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: receiverUid
        });
        Object.defineProperty(this, "metaId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: metaId
        });
    }
    static fromJSON(metaBody) {
        if (metaBody instanceof MetaBody === false) {
            metaBody = new MetaBody(metaBody.type, metaBody.senderUid, metaBody.data, metaBody.streamId, metaBody.receiverUid, metaBody.metaId);
        }
        return metaBody;
    }
    static fromText(senderUid, data, streamId, receiverUid) {
        return new MetaBody(streamId == null
            ? IPC_META_BODY_TYPE.INLINE_TEXT
            : IPC_META_BODY_TYPE.STREAM_WITH_TEXT, senderUid, data, streamId, receiverUid);
    }
    static fromBase64(senderUid, data, streamId, receiverUid) {
        return new MetaBody(streamId == null
            ? IPC_META_BODY_TYPE.INLINE_BASE64
            : IPC_META_BODY_TYPE.STREAM_WITH_BASE64, senderUid, data, streamId, receiverUid);
    }
    static fromBinary(sender, data, streamId, receiverUid) {
        if (typeof sender === "number") {
            return new MetaBody(streamId == null
                ? IPC_META_BODY_TYPE.INLINE_BINARY
                : IPC_META_BODY_TYPE.STREAM_WITH_BINARY, sender, data, streamId, receiverUid);
        }
        if (sender.support_binary) {
            return this.fromBinary(sender.uid, data, streamId, receiverUid);
        }
        return this.fromBase64(sender.uid, simpleDecoder(data, "base64"), streamId, receiverUid);
    }
    get type_encoding() {
        const encoding = this.type & 0b11111110;
        switch (encoding) {
            case IPC_DATA_ENCODING.UTF8:
                return IPC_DATA_ENCODING.UTF8;
            case IPC_DATA_ENCODING.BASE64:
                return IPC_DATA_ENCODING.BASE64;
            case IPC_DATA_ENCODING.BINARY:
                return IPC_DATA_ENCODING.BINARY;
        }
    }
    get type_isInline() {
        return (this.type & IPC_META_BODY_TYPE.INLINE) !== 0;
    }
    get type_isStream() {
        return (this.type & IPC_META_BODY_TYPE.INLINE) === 0;
    }
    get jsonAble() {
        if (this.type_encoding === IPC_DATA_ENCODING.BINARY) {
            return MetaBody.fromBase64(this.senderUid, simpleDecoder(this.data, "base64"), this.streamId, this.receiverUid);
        }
        return this;
    }
    toJSON() {
        return { ...this.jsonAble };
    }
}
__decorate([
    cacheGetter()
], MetaBody.prototype, "type_encoding", null);
__decorate([
    cacheGetter()
], MetaBody.prototype, "type_isInline", null);
__decorate([
    cacheGetter()
], MetaBody.prototype, "type_isStream", null);
__decorate([
    cacheGetter()
], MetaBody.prototype, "jsonAble", null);
export var IPC_META_BODY_TYPE;
(function (IPC_META_BODY_TYPE) {
    /** 流 */
    IPC_META_BODY_TYPE[IPC_META_BODY_TYPE["STREAM_ID"] = 0] = "STREAM_ID";
    /** 内联数据 */
    IPC_META_BODY_TYPE[IPC_META_BODY_TYPE["INLINE"] = 1] = "INLINE";
    /** 流，但是携带一帧的 UTF8 数据 */
    IPC_META_BODY_TYPE[IPC_META_BODY_TYPE["STREAM_WITH_TEXT"] = 2] = "STREAM_WITH_TEXT";
    /** 流，但是携带一帧的 BASE64 数据 */
    IPC_META_BODY_TYPE[IPC_META_BODY_TYPE["STREAM_WITH_BASE64"] = 4] = "STREAM_WITH_BASE64";
    /** 流，但是携带一帧的 BINARY 数据 */
    IPC_META_BODY_TYPE[IPC_META_BODY_TYPE["STREAM_WITH_BINARY"] = 8] = "STREAM_WITH_BINARY";
    /** 内联 UTF8 数据 */
    IPC_META_BODY_TYPE[IPC_META_BODY_TYPE["INLINE_TEXT"] = 3] = "INLINE_TEXT";
    /** 内联 BASE64 数据 */
    IPC_META_BODY_TYPE[IPC_META_BODY_TYPE["INLINE_BASE64"] = 5] = "INLINE_BASE64";
    /** 内联 BINARY 数据 */
    IPC_META_BODY_TYPE[IPC_META_BODY_TYPE["INLINE_BINARY"] = 9] = "INLINE_BINARY";
})(IPC_META_BODY_TYPE || (IPC_META_BODY_TYPE = {}));

File: android/MicroService/src/main/java/org/dweb_browser/microservice/ipc/message/MetaBody.kt
package org.dweb_browser.microservice.ipc.message

import com.google.gson.*
import com.google.gson.annotations.JsonAdapter
import org.dweb_browser.helper.toBase64
import org.dweb_browser.helper.toBase64Url
import org.dweb_browser.microservice.ipc.Ipc
import java.lang.reflect.Type
import java.util.*

@JsonAdapter(MetaBody::class)
data class MetaBody(
  /**
   * 类型信息，包含了 编码信息 与 形态信息
   * 编码信息是对 data 的解释
   * 形态信息（流、内联）是对 "是否启用 streamId" 的描述（注意，流也可以内联第一帧的数据）
   */
  val type: IPC_META_BODY_TYPE,
  val senderUid: Int,
  val data: Any,
  val streamId: String? = null,
  var receiverUid: Int? = null,
  /**
   * 唯一id，指代这个数据的句柄
   *
   * 需要使用这个值对应的数据进行缓存操作
   * 远端可以发送句柄回来，这样可以省去一些数据的回传延迟。
   */
  val metaId: String = ByteArray(8).also { Random().nextBytes(it) }.toBase64Url()
) : JsonSerializer<MetaBody>, JsonDeserializer<MetaBody> {

  @JsonAdapter(IPC_META_BODY_TYPE::class)
  enum class IPC_META_BODY_TYPE(val type: Int) : JsonSerializer<IPC_META_BODY_TYPE>,
    JsonDeserializer<IPC_META_BODY_TYPE> {
    /** 流 */
    STREAM_ID(0),

    /** 内联数据 */
    INLINE(1),


    /** 文本 json html 等 */
    STREAM_WITH_TEXT(STREAM_ID or IPC_DATA_ENCODING.UTF8),

    /** 使用文本表示的二进制 */
    STREAM_WITH_BASE64(STREAM_ID or IPC_DATA_ENCODING.BASE64),

    /** 二进制 */
    STREAM_WITH_BINARY(STREAM_ID or IPC_DATA_ENCODING.BINARY),

    /** 文本 json html 等 */
    INLINE_TEXT(INLINE or IPC_DATA_ENCODING.UTF8),

    /** 使用文本表示的二进制 */
    INLINE_BASE64(INLINE or IPC_DATA_ENCODING.BASE64),

    /** 二进制 */
    INLINE_BINARY(INLINE or IPC_DATA_ENCODING.BINARY), ;

    val encoding by lazy {
      val encoding = type and 0b11111110;
      IPC_DATA_ENCODING.values().find { it.encoding == encoding }
    }
    val isInline by lazy {
      type and 1 == 1
    }
    val isStream by lazy {
      type and 1 == 0
    }

    override fun serialize(
      src: IPC_META_BODY_TYPE, typeOfSrc: Type?, context: JsonSerializationContext?
    ) = JsonPrimitive(src.type)

    override fun deserialize(
      json: JsonElement, typeOfT: Type?, context: JsonDeserializationContext?
    ) = json.asInt.let { type -> values().find { it.type == type } }

    private inline infix fun or(TYPE: IPC_DATA_ENCODING) = type or TYPE.encoding
  }


  companion object {
    fun fromText(
      senderUid: Int,
      data: String,
      streamId: String? = null,
      receiverUid: Int? = null,
    ) = MetaBody(
      type = if (streamId == null) IPC_META_BODY_TYPE.INLINE_TEXT else IPC_META_BODY_TYPE.STREAM_WITH_TEXT,
      senderUid = senderUid,
      data = data,
      streamId = streamId,
      receiverUid = receiverUid,
    )

    fun fromBase64(
      senderUid: Int,
      data: String,
      streamId: String? = null,
      receiverUid: Int? = null,
    ) = MetaBody(
      type = if (streamId == null) IPC_META_BODY_TYPE.INLINE_BASE64 else IPC_META_BODY_TYPE.STREAM_WITH_BASE64,
      senderUid = senderUid,
      data = data,
      streamId = streamId,
      receiverUid = receiverUid,
    )

    fun fromBinary(
      senderUid: Int,
      data: ByteArray,
      streamId: String? = null,
      receiverUid: Int? = null,
    ) = MetaBody(
      type = if (streamId == null) IPC_META_BODY_TYPE.INLINE_BINARY else IPC_META_BODY_TYPE.STREAM_WITH_BINARY,
      senderUid = senderUid,
      data = data,
      streamId = streamId,
      receiverUid = receiverUid,
    )

    fun fromBinary(
      senderIpc: Ipc,
      data: ByteArray,
      streamId: String? = null,
      receiverUid: Int? = null,
    ) = if (senderIpc.supportBinary) fromBinary(
      senderIpc.uid, data, streamId, receiverUid
    ) else fromBase64(
      senderIpc.uid, data.toBase64(), streamId, receiverUid
    )
  }


  val jsonAble by lazy {
    when (type.encoding) {
      IPC_DATA_ENCODING.BINARY -> fromBase64(
        senderUid,
        (data as ByteArray).toBase64(),
        streamId,
        receiverUid
      )

      else -> this
    }
  }

  override fun serialize(
    src: MetaBody, typeOfSrc: Type, context: JsonSerializationContext
  ) = JsonObject().also { jsonObject ->
    with(src.jsonAble) {
      jsonObject.add("type", context.serialize(type))
      jsonObject.addProperty("senderUid", senderUid)
      jsonObject.addProperty("data", data as String)
      jsonObject.addProperty("streamId", streamId)
      jsonObject.addProperty("receiverUid", receiverUid)
      jsonObject.addProperty("metaId", metaId)
    }
  }


  override fun deserialize(
    json: JsonElement,
    typeOfT: Type,
    context: JsonDeserializationContext
  ): MetaBody = json.asJsonObject.let { obj ->
    MetaBody(
      type = context.deserialize(obj["type"], IPC_META_BODY_TYPE::class.java),
      senderUid = obj["senderUid"].asInt,
      data = obj["data"].asString,
      streamId = obj.getStringOrNull("streamId"),
      receiverUid = obj.getIntOrNull("receiverUid"),
      metaId = obj["metaId"].asString,
    )
  }
}

private fun JsonObject.getElementOrNull(key: String): JsonElement? {
  if (!has(key)) {
    return null
  }
  val value = get(key)
  return if (value.isJsonNull)
    null
  else value
}

private fun JsonObject.getStringOrNull(key: String) = getElementOrNull(key)?.asString

private fun JsonObject.getIntOrNull(key: String) = getElementOrNull(key)?.asInt


